proposals/index-range-changes.md

```diff
@@ -0,0 +1,297 @@
+- The type does not have an instance indexer which takes a single `Range` as the parameter.
+
+For such types, the language will bind as if there is an index member of the form `T this[Range range]` where `T` is 
+the return type of the `Slice` method including any `ref` style annotations. The new member will also have matching
+accessibility with `Slice`. 
+
+When the `Range` based indexer is bound on an expression named `receiver`, it will be lowered by converting the `Range` 
+expression into two values that are then passed to the `Slice` method. For discussion purposes, lets use the example of 
+`receiver[expr]`.
+
+The first argument of `Slice` will be obtained by converting the typed expression in the following way:
+
+- When `expr` is of the form `expr1..expr2` (where `expr2` can be omitted) and `expr1` has type `int`, then it will be
+emitted as `expr1`.
+- When `expr` is of the form `^expr1..expr2` (where `expr2` can be omitted), then it will be emitted as 
+`receiver.Length - expr1`.
+- When `expr` is of the form `..expr2` (where `expr2` can be omitted), then it will be emitted as `0`.
+- Otherwise, it will be emitted as `expr.Start.GetOffset(receiver.Length)`.
+
+This value will be re-used in the calculation of the second `Slice` argument. When doing so it will be referred to as
+`start`. The second argument of `Slice` will be obtained by converting the range typed expression in the following way:
+
+- When `expr` is of the form `expr1..expr2` (where `expr1` can be omitted) and `expr2` has type `int`, then it will be 
+emitted as `expr2 - start`.
+- When `expr` is of the form `expr1..^expr2` (where `expr1` can be omitted), then it will be emitted as 
+`(receiver.Length - expr2) - start`.
+- When `expr` is of the form `expr1..` (where `expr1` can be omitted), then it will be emitted as `receiver.Length`.
+- Otherwise, it will be emitted as `expr.End.GetOffset(receiver.Length) - start`.
+
+The `receiver`, `Length` and `expr` expressions will be spilled as appropriate to ensure any side effects are only
+executed once. For example:
+
+``` csharp
+class Collection {
+    private int[] _array = new[] { 1, 2, 3 };
+
+    int Length {
+        get {
+            Console.Write("Length ");
+            return _array.Length;
+        }
+    }
+
+    int[] Slice(int start, int length) { 
+        var slice = new int[length];
+        Array.Copy(_array, start, slice, 0, length);
+        return slice;
+    }
+}
+
+class SideEffect {
+    Collection Get() {
+        Console.Write("Get ");
+        return new Collection();
+    }
+
+    void Use() { 
+        var array = Get()[0..2];
+        Console.WriteLine(array.length);
+    }
+}
+```
+
+This code will print "Get Length 2". 
+
+The language will special case the following known types: 
+
+- `string`: the method `Substring` will be used instead of `Slice`.
+- `array`: the method `System.Reflection.CompilerServices.GetSubArray` will be used instead of `Slice`.
```

Comment:
Nit: System.Reflection.CompilerServices.GetSubArray => System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray

It's not a big deal, but I also still question why we defined this method on RuntimeHelpers rather than just making it a static method on Array, along with all the other static generic helper methods on Array.  Seems arbitrary to bury it here when a developer could reasonably call it directly.
cc: @tarekgh, @terrajobst

Reply:
I think having it in System.Runtime.CompilerServices.RuntimeHelpers is implicitly saying only the compiler suppose to call it. why you think the developers will need to call it if they can get the same result in other way?

Reply:
> why you think the developers will need to call it if they can get the same result in other way?

Does VB expose this range-based array slicing functionality?  F#?

Reply:
I am not aware of any support of VB or F# for array slicing, @jaredpar may know if there is any work there. But if they do, wouldn't they use System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray too?

Reply:
That's my attempt at a point :)  Assuming there is no built-in language support, it'd be nice if F# and VB could just call Array.Slice rather than calling System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray.

Reply:
That make sense. @agocke replied on the other issue that change will be a real breaking change. Earlier version of Roslyn will fail to compile against the newer framework, and vice versa

Reply:
@cartermp for F# discussion